### PR TITLE
Remember last written LSN when it is first requested

### PR DESCRIPTION
**Pull request by knizhnik**
See https://neondb.slack.com/archives/C03QLRH7PPD/p1712529369520409

In case of statements CREATE TABLE AS SELECT... or INSERT FROM SELECT... we are fetching data from source table and storing it in destination table. It cause problems with prefetch last-written-lsn is known for the pages of source table
(which for example happens after compute restart). In this case we get get global value of last-written-lsn which is changed frequently as far as we are writing pages of destination table. As a result request-isn for the prefetch and request-let when this page is actually needed are different and we got exported prefetch request. So it actually disarms prefetch.

Proposed simple patch stores last-written LSN for the page when it is not found. So next time we will request last-written LSN for this page, we will get the same value (certainly if the page was not changed).

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -9038,6 +9038,12 @@ GetLastWrittenLSN(RelFileNode rnode, ForkNumber forknum, BlockNumber blkno)
 		entry = hash_search(lastWrittenLsnCache, &key, HASH_FIND, NULL);
 		if (entry != NULL)
 			lsn = entry->lsn;
+		else
+		{
+			LWLockRelease(LastWrittenLsnLock);
+			SetLastWrittenLSNForBlock(lsn, rnode, forknum, blkno);
+			return lsn;
+		}
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply src/backend/access/transam/xlog.c:9046]: > This doesn't guarantee correctness, as we may actually fail to see concurrent updates to the LSN .

Sorry, can you explain how we can fail to see concurrent updates?
Please notice, that `SetLastWrittenLSNForBlock` preforms check and replace existed value only if it is smaller than new one.

[the remark this answers is not in the document]
Comment (knizhnik) [reply src/backend/access/transam/xlog.c:9046]: Sorry, but I still do not see the problem.
Can you provide some schedule which demonstrate it?
So we have two backends: first one (B1) performs prefetch of page P, it calls GetLastWrittenLsn(P), it is not found and we take some global last-written LSN (L1).
Another backend B2 tries to update the same page P and set last-written LSN L2>L1 for it.
Possible schedules:

First:
P1: SetLastWrittenLsn(L1)    =>  LWLSN(P)=L1
P2: SetLastWrittenLsn(L2)   => LWLSN(P)=L2
P1: prefetch page with LSN=L1
P1: read page with LSN=GetLastWrittenLSN(P)=L2 -> prefetch request expired

Another:
P2:  SetLastWrittenLsn(L2)   => LWLSN(P)=L2
P1:  SetLastWrittenLsn(L1)    =>  LWLSN(P)=L2
P1: prefetch page with LSN=L1
P1: read page with LSN=GetLastWrittenLSN(P)=L2 -> prefetch request expired

In which case we can get the problem?
Please look at this problem from the other angle. SetLastWrittenLSN never moved LSN backward - it can only increase it. So if we do not call SetLastWrittenLSN as it is suggested in this PR, then it keeps old smaller value. But it is always safe to overestimate (use larger value) of LWLSN. Otherwise our LWLSN cache with LRU replacement algorithm will too work at all.

In other words extra call of SetLastWrittenLSN can not somehow affect  correctness of this algorithm.

[the remark this answers is not in the document]
Comment (knizhnik) [reply src/backend/access/transam/xlog.c]: Frankly speaking I do not think that this scenario is very important: the case when same relation is fetched and updated at the same time seems to be quite rare. More precisely: certainly vacuum and just seqscan which sets hint bits actually doing it: update fetch pages. But it never tries to update page which is not yet fetched:)

But I agree with your statement: want to prevent already-expired LwLSNs as much as possible.
And I see not reasons why we could not slightly change `GetLastWrittenLSN` to return most recent value if it is possible. The only thing which is needed for it - let `SetLastWrittenLSN ` return updated value. 
I committed updated version of the patch. 
Does it address your concerns now?
